### PR TITLE
Re-add Functor/Applicative constraints for GHC 7.8 compatibility

### PR DESCRIPTION
**Pull request by charleso**
@brendanhay Just starting to looking to upgrade our version of amazonka and this is the first thing we hit.

We've hit some other snags around the transformers, but that's a problem for another day...

### CODE REVIEW

**Comment by brendanhay** [comment]
Wow! I didn't even know it built on `6.8` :smile:

**Comment by brendanhay** [comment]
If this is a supported use case, I'll some `6.8` buildsto the `travis.yml`.

**Comment by charleso** [comment]
@brendanhay DOH! I mean 7.8. I'm an idiot...

**Comment by charleso** [comment]
Now I'm wondering how the travis builds are passing?

https://travis-ci.org/brendanhay/amazonka/jobs/72418403

**Comment by brendanhay** [comment]
Aha, good catch. Could it be that since `amazonka` proper doesn't have a `test-suite` declaration in the cabal file, it's not actually being built on CI due to `stack setup/stack test`? If so, that's an explosively wet brain fart on my part.

**Comment by charleso** [comment]
Hmm. I couldn't say. You can see 'build' in the build output above, which suggests to me it is doing _something_

**Comment by brendanhay** [comment]
I'll spend some time today verifying it fails _without_ your PR, before merging.

**Comment by charleso** [comment]
No worries, and thanks. I'm just trying to build in isolation with stack.

**Comment by brendanhay** [comment]
So it indeed turned out to be the lack of a `test-suite` in the `amazonka.cabal` configuration. By adding a basic `test-suite`, the build has started erroring: https://travis-ci.org/brendanhay/amazonka/jobs/72423722

Thanks for bringing the falsch CI pancea to my attention.

File: amazonka/src/Control/Monad/Error/AWS.hs [hunk 1]
```diff
@@ -69,11 +69,12 @@ runAWST :: (MonadCatch m, MonadResource m, AWSEnv r)
 runAWST e = runExceptT . AWST.runAWST e
 {-# DEPRECATED runAWST "Exists for backwards compatibility with pre-@1.0@." #-}
 
-hoistError :: (MonadError e m, AWSError e) => Either Error a -> m a
+hoistError :: (Applicative m, MonadError e m, AWSError e) => Either Error a -> m a
 hoistError = either (throwing _Error) pure
```
Comment (brendanhay) [inline amazonka/src/Control/Monad/Error/AWS.hs]: Would changing the use of `pure` to `return` remove the need for the `Applicative` constraint entirely?

Comment (charleso) [inline amazonka/src/Control/Monad/Error/AWS.hs]: > Would changing the use of pure to return remove the need for the Applicative constraint entirely?

By the looks of it - yes. Sorry, I wasn't paying attention. That's a good idea.

**Comment by brendanhay** [comment]
I've merged and made the `pure` -> `return` switch in `develop`.

**Comment by charleso** [comment]
Thanks @brendanhay

**Pull request merged**